### ISSUE TITLE
Add "delete" function to files

### ISSUE THREAD

**Issue by frenzluzern** [HubSpot/hubspot-cms-deploy-action#10]
Hi, i would like to have a file deleted in HubSpot if it is deleted on the GitHub repository. Right now the files are just updated and uploaded.  Thank you so much.

**Comment by TheWebTech** [comment]
The idea makes sense. May be tricky to implement.

Right now the GH action basically uses the HubSpot CLI to run the upload command, uploading all of the files in your selected branch.

When the action runs, the only version of the files it has is just what is already committed. It doesn't know what was deleted. Not exactly sure how we can safely execute the removal of deleted files.

**Comment by TheWebTech** [comment]
@cartwriting **if you have pages, modules or templates that depend on the asset that's being deleted how would you want us to handle that scenario?**  Removing the asset would break all of the assets depending on it.

Should it just fail to deploy? Would you notice the deployment failure or would you think the GH action was broken?

---


Regarding the documentation instructing to lock the folder in the Design Manager - you can still delete files via the CLI. `hs remove` does this.

**Comment by neilyonzon1** [comment]
@TheWebTech - We would not want to allow a file or folder to be deleted if it will break assets that depend on it. I know with the current action behavior under the 'Deploy' step we can see a log of failed uploads. Something similar for 'failed deletions due to dependencies' would be great as that would prompt us to correct/reroute asset dependencies.

Additionally, If there was a more overt way of catching these errors without going inside the Deploy step log would be great.

**Comment by TheWebTech** [comment]
Should the full upload fail or just the individual assets that have dependencies?

I ask because say you created new templates calling said modules or resources, they may now be new dependents.

If we fail the whole build it prevents that broken scenario.

**Comment by neilyonzon1** [comment]
@TheWebTech - Makes sense to have the full upload to fail in case there were brand new dependents. Perhaps if there was a way to hold off deletions at a separate step after the upload, the action would recognize the new dependencies and error the deletions?

**Comment by johnsfuller** [comment]
YES PLEASE ADD THIS!!! Sorry I yelled.

**Comment by code3z** [comment]
yes! Also, husbpot needs to detect when I move a module based on ID and update the dependencies accordingly, just like it does in the web UI. A github integration that requires unlocking the design manager and messing with things doesn't make sense - github is supposed to be the source of truth

**Comment by daniel-layers** [comment]
Strangely, this is still not remedied 2 years later, I'm just getting things set for the GitHub actions picking up pushes for sandbox and pushes for PRD but it's quite concerning that the files that are removed in the Git commit are not removed from the files in HubSpot. Especially when using GitHub/Version control so that you have one source of truth is the purpose for this.

**Comment by TanyaScales** [comment]
Hello to everyone on this thread -- Thank you for all your input -- the issue here with the removal of files is definitely a complex one from the perspective of dependencies, CI/CD, and the error messaging and mitigation that is being discussed. However, I definitely understand the value of having those files that you've determined are no longer necessary also be removed from the file system in the Design Manager. There are a few open issues in the CLI and its library that would work towards making this more of a reality over here in the GitHub Action.

Question: If a removed file has dependencies in the HubSpot portal/Design Manager -- and during upload, those were listed for you as being unable to be removed. What is the information you would be looking for in order to action those? (Link to the file in the DM? # of dependencies found on that file?) -- Or conversely, would you be fine that it alerted you to the reason those files weren't removed on upload and move on without addressing them?

**Comment by daniel-layers** [comment]
@TanyaScales if I'm pushing my Git and deleting a file it's because it's no longer used and 9/10s out of 10 is an image asset or similar that is now redundant. 

I personally would be fine with it removing them and a note in the completed action of files that had dependencies that were removed. I wouldn't want it to reject the deletion of them as the aim of using Git with HubSpot is to have 1 version of truth it's not possible to do that currently because HubSpot won't remove the files therefore there is no truth only 2 variations of versions.

**Comment by TheWebTech** [comment]
Daniel, what Tanya is for example, we know you have pages using a template or module. You delete the template or module in your repo but then when syncing the change to HubSpot on HubSpots end we have to prevent the deletion from going through because if we don't you will instantly have broken pages on your website.

You need to delete the references in your pages if you don't want it anymore, delete pages, or swap templates. Otherwise you've got broken pages.

**Comment by daniel-layers** [comment]
@TheWebTech I am aware of what the original commentor meant and to me it's fine if they're deleted and a note in the action completed log saying XYZ was deleted with dependencies in: and lists the files that had dependencies. 

To me you shouldn't be deleting things you're unsure of and you should check dependencies before make sure your commits are streamlined and exact, not "let me delete all this and see what bounces back" that's not how Git should be used.

**Comment by ranjithvkechidna** [comment]
Can't we have the "--clean" option of the "hs upload" command used here to ensure the files are deleted that are deleted from Git?
<img width="1467" alt="image" src="https://github.com/HubSpot/hubspot-cms-deploy-action/assets/25499885/93f62606-8582-4a6d-a89a-8d0dbc113554">

Otherwise, it's a lot of manual work to delete and clean up the unnecessary files everywhere starting locally with Git, the HubSpot Sandbox as well as the Production accounts

**Comment by ranjithvkechidna** [comment]
> @TheWebTech I am aware of what the original commentor meant and to me it's fine if they're deleted and a note in the action completed log saying XYZ was deleted with dependencies in: and lists the files that had dependencies.
> 
> To me you shouldn't be deleting things you're unsure of and you should check dependencies before make sure your commits are streamlined and exact, not "let me delete all this and see what bounces back" that's not how Git should be used.

I agree with this and as long as the files were part of the git and not uploaded directly through the editor (which defeats the purpose of maintaining git) we should be anytime able to recover it back if the delete action was an accident.

**Comment by TheWebTech** [comment]
> Can't we have the "--clean" option of the "hs upload" command used here to ensure the files are deleted that are deleted from Git?
> <img width="1467" alt="image" src="https://github.com/HubSpot/hubspot-cms-deploy-action/assets/25499885/93f62606-8582-4a6d-a89a-8d0dbc113554">
> 
> Otherwise, it's a lot of manual work to delete and clean up the unnecessary files everywhere starting locally with Git, the HubSpot Sandbox as well as the Production accounts

The clean argument was implemented partially to solve for this.

Currently we do not have logging or dependency checking going on with this. It simply deletes anything in the directory that's not what's being uploaded.

> I agree with this and as long as the files were part of the git and not uploaded directly through the editor (which defeats the purpose of maintaining git) we should be anytime able to recover it back if the delete action was an accident.

You may not be able to recover back is the issue.  
Pages for example using a deleted module, not hard coded into the template would lose the module. Reverting to a prior commit that has the module may not restore those module instances especially if the page was edited after losing the module. If you have templates that load the module via id instead of path, the id of the reverted module may be different from the original resulting in pages that had that module not being able to tie the  editor saved values to the reverted version of the module 

If you delete templates your pages using that template would likely begin throwing `500 errors` because the template doesn't exist. You would not see any warnings or errors from the GH action. I'm not sure right now how the page editor will handle that. I do think in this case restoring a template file likely will fix the pages using it though. I think that would be a more acceptable situation than with modules.

Reusable section templates, and theme settings/config are also possible areas where you may be able to put yourself into a broken state that is otherwise not possible, and recovering is then not as simple as reverting to a prior commit.

So to sum it up: you could potentially break your site in a way that you can't easily revert/fix and HubSpot Support may also have trouble fixing. 

Will bring it to the team though. If in testing we find the degradation and your ability to revert acceptable then it makes it feasible to push this forward. If not then we'll need to continue discussing both internally and with y'all to strike the right solution/balance.
